apps/labs/posts/scaling-asyncio-on-free-threaded-python.md

```diff
@@ -0,0 +1,177 @@
+multiple threads. In the general case of multiple event loops running in
+parallel, there is no lock contention and scales linearly with the number of
+threads.
+
+Here are the key changes:
+
+1. **Per-thread linked list of tasks**:
+   Python 3.14 introduces a per-thread circular double linked list implementation
+   for storing tasks instead of a global `WeakSet`. The linked list is per-thread,
+   meaning that each thread maintains its own list of tasks and allows for
+   lock-free addition and removal of tasks. Weak references are slow and
+   and prone to contention. The new implementation removes the use of weak
+   references entirely and makes tasks responsible for removing themselves
+   from the list when they are done.
+   This requires cooperation between the task's deallocator and the executing
+   threads to ensure that the task is removed from the list before it is freed,
+   otherwise a thread could try accessing an already freed task. By removing the
+   use of weak references, the overhead of reference counting is eliminated
+   entirely and addition/removal of a task in the list now requires only updating
+   the pointers in the linked list.
+
+   This design allows for efficient, lock-free and thread-safe task management
+   and scales well on the free-threaded interpreter.
+
+   This was implemented in https://github.com/python/cpython/pull/128869.
+
+2. **Per-thread current task**:
+   Python 3.14 stores the current task in the thread state structure,
+   which is local to each thread. By storing the current task on the thread state,
+   the overhead of accessing the current task is reduced, allowing for lock-free
+   access to the current task while avoiding dictionary lookup.
+   This allows for faster switching between tasks -- a very frequent operation in asyncio.
+
+   This design allows for lock-free access to the current task and avoids
+   reference counting and lock contention on the global dictionary.
+
+   This was implemented in https://github.com/python/cpython/pull/129899.
+
+Both of these changes allow `asyncio` to scale linearly with the number of
+threads in free-threading, and has significantly improved performance for both
+single-threaded and multi-threaded `asyncio` usage. The standard `pyperformance`
+benchmark suite shows a significant 10–20% improvement in single-threaded performance while also
+reducing memory usage.
+
+For a deeper dive into the implementation, check out the internal docs for
+[asyncio](https://github.com/python/cpython/blob/main/InternalDocs/asyncio.md).
+
+
+## Benchmarks
+
+Here are the benchmark results comparing the performance of `asyncio` on the free-threaded
+build with the GIL-enabled build on a Windows machine with 6 physical CPU cores and 12 hyper-threads:
+
+- **TCP Benchmark**: This benchmark measures raw TCP performance.
+
+  <figure style={{ textAlign: 'center' }}>
+    <img
+      src="/posts/scaling-asyncio-on-free-threaded-python/asyncio_tcp_benchmark.png"
```

Comment:
The green numbers are really hard to read, and red/green is the most common color combo that cannot be distinguished by people who are color blind. If it's not too much trouble, it'd be nice to regenerate this with better colors. Replacing green with dark blue for example.

Reply:
Much nicer, thanks